src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,464 @@
+            for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+                if (gpu_freq_request.at(domain_idx) !=
+                    m_gpu_freq_control.at(domain_idx).last_setting) {
+                    m_platform_io.adjust(m_gpu_freq_control.at(domain_idx).batch_idx,
+                                         gpu_freq_request.at(domain_idx));
+                    m_gpu_freq_control.at(domain_idx).last_setting =
+                                         gpu_freq_request.at(domain_idx);
+                    ++m_gpu_frequency_requests;
+                }
+            }
+            m_do_write_batch = true;
+        }
+    }
+
+    // If controls have a valid updated value write them.
+    bool GPUActivityAgent::do_write_batch(void) const
+    {
+        return m_do_write_batch;
+    }
+
+    // Read signals from the platform and calculate samples to be sent up
+    void GPUActivityAgent::sample_platform(std::vector<double> &out_sample)
+    {
+        assert(out_sample.size() == M_NUM_SAMPLE);
+
+        // Collect latest signal values
+        for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
+            m_gpu_freq_status.at(domain_idx).value = m_platform_io.sample(m_gpu_freq_status.at(
+                                                                          domain_idx).batch_idx);
+            m_gpu_compute_activity.at(domain_idx).value = m_platform_io.sample(m_gpu_compute_activity.at(
+                                                                               domain_idx).batch_idx);
+            m_gpu_utilization.at(domain_idx).value = m_platform_io.sample(m_gpu_utilization.at(
+                                                                          domain_idx).batch_idx);
+
+            m_gpu_energy.at(domain_idx).value = m_platform_io.sample(m_gpu_energy.at(
+                                                                     domain_idx).batch_idx);
+        }
+
+        m_time.value = m_platform_io.sample(m_time.batch_idx);
+    }
+
+    // Wait for the remaining cycle time to keep Controller loop cadence
+    void GPUActivityAgent::wait(void)
+    {
+        geopm_time_s current_time;
+        do {
+            geopm_time(&current_time);
+        }
+        while(geopm_time_diff(&m_last_wait, &current_time) < M_WAIT_SEC);
+        geopm_time(&m_last_wait);
+    }
+
+    // Adds the wait time to the top of the report
+    std::vector<std::pair<std::string, std::string> > GPUActivityAgent::report_header(void) const
+    {
+        return {{"Wait time (sec)", std::to_string(M_WAIT_SEC)}};
+    }
+
+    // Adds number of frquency requests to the per-node section of the report
+    std::vector<std::pair<std::string, std::string> > GPUActivityAgent::report_host(void) const
+    {
+        std::vector<std::pair<std::string, std::string> > result;
+
+        result.push_back({"GPU Frequency Requests", std::to_string(m_gpu_frequency_requests)});
+#ifdef GEOPM_DEBUG
```

Comment:
Using a debug build is going to slow down the overall runtime.  I'm not sure how big the impact would be on this Agent without doing some testing. It seems that if you're trying to study something around clipping, you ought to use a release build to ensure things run at full speed.

Why not just include this all the time?

Reply:
As long as it doesn't add too much to the report including it all the time seems reasonable.